compiler/luci/service/src/Nodes/CircleReshape.test.cpp

```diff
@@ -37,3 +38,241 @@ TEST(CloneNodeTest, clone_Reshape)
+  ASSERT_TRUE(shape.dim(1).known());
+  ASSERT_EQ(6, shape.dim(0).value());
+  ASSERT_EQ(4, shape.dim(1).value());
+}
+
+TEST(ShapeRuleTest, reshape_output_unknown_dim)
+{
+  luci::CircleInput input;
+  input.shape({2, 3, 4});
+  input.shape_status(luci::ShapeStatus::VALID);
+
+  luci::CircleConst new_shape;
+  new_shape.dtype(loco::DataType::S32);
+  new_shape.size<loco::DataType::S32>(2);
+  new_shape.at<loco::DataType::S32>(0) = -1;
+  new_shape.at<loco::DataType::S32>(1) = 4;
+  new_shape.shape_status(luci::ShapeStatus::VALID);
+
+  luci::CircleReshape reshape;
+  reshape.tensor(&input);
+  reshape.shape(&new_shape);
+
+  loco::TensorShape shape;
+  luci::sinf::Rule shape_inf_rule;
+
+  ASSERT_TRUE(shape_inf_rule.infer(&reshape, shape));
+
+  ASSERT_EQ(3, input.rank());
+  ASSERT_TRUE(input.dim(0).known());
+  ASSERT_TRUE(input.dim(1).known());
+  ASSERT_TRUE(input.dim(2).known());
+  ASSERT_EQ(2, input.dim(0).value());
+  ASSERT_EQ(3, input.dim(1).value());
+  ASSERT_EQ(4, input.dim(2).value());
+  ASSERT_EQ(2, shape.rank());
+  ASSERT_TRUE(shape.dim(0).known());
+  ASSERT_TRUE(shape.dim(1).known());
+  ASSERT_EQ(6, shape.dim(0).value());
+  ASSERT_EQ(4, shape.dim(1).value());
+}
+
+TEST(ShapeRuleTest, reshape_both_unknown_dim)
+{
+  luci::CircleInput input;
+  input.shape({2, 3, 4});
+  input.dim(0).unset();
+  input.shape_status(luci::ShapeStatus::VALID);
+
+  luci::CircleConst new_shape;
+  new_shape.dtype(loco::DataType::S32);
+  new_shape.size<loco::DataType::S32>(2);
+  new_shape.at<loco::DataType::S32>(0) = -1;
+  new_shape.at<loco::DataType::S32>(1) = 4;
+  new_shape.shape_status(luci::ShapeStatus::VALID);
+
+  luci::CircleReshape reshape;
+  reshape.tensor(&input);
+  reshape.shape(&new_shape);
+
+  loco::TensorShape shape;
+  luci::sinf::Rule shape_inf_rule;
+
+  ASSERT_TRUE(shape_inf_rule.infer(&reshape, shape));
+
+  ASSERT_EQ(3, input.rank());
+  ASSERT_FALSE(input.dim(0).known());
+  ASSERT_TRUE(input.dim(1).known());
+  ASSERT_TRUE(input.dim(2).known());
+  ASSERT_EQ(0, input.dim(0).value());
+  ASSERT_EQ(3, input.dim(1).value());
+  ASSERT_EQ(4, input.dim(2).value());
+  ASSERT_EQ(2, shape.rank());
+  ASSERT_FALSE(shape.dim(0).known());
+  ASSERT_TRUE(shape.dim(1).known());
+  ASSERT_EQ(0, shape.dim(0).value());
+  ASSERT_EQ(4, shape.dim(1).value());
+}
+
+TEST(ShapeRuleTest, reshape_output_multiple_unknown_dim_1_NEG)
+{
+  luci::CircleInput input;
+  input.shape({2, 3, 4});
+  input.shape_status(luci::ShapeStatus::VALID);
+
+  luci::CircleConst new_shape;
+  new_shape.dtype(loco::DataType::S32);
+  new_shape.size<loco::DataType::S32>(2);
+  new_shape.at<loco::DataType::S32>(0) = -1;
+  new_shape.at<loco::DataType::S32>(1) = -1;
+  new_shape.shape_status(luci::ShapeStatus::VALID);
+
+  luci::CircleReshape reshape;
+  reshape.tensor(&input);
+  reshape.shape(&new_shape);
+
+  loco::TensorShape shape;
+  luci::sinf::Rule shape_inf_rule;
+
+  ASSERT_ANY_THROW(shape_inf_rule.infer(&reshape, shape));
```

Comment:
why is it thrown?

Reply:
I threw an exception when the number of unknown dimensions is more than 1 in new shape.
I follow the behavior of the tensorflow lite's, and I thought it could be reasonable because constant new shape can be inferred only up for 1 dimension.
Tensorflow lite throws following error in that case: `error: 'tf.Reshape' op requires 'shape' to have at most one dynamic dimension, but got multiple dynamic dimensions at indices 0 and 1`